pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,237 @@
+package sa
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/serviceaccounts"
+	osautil "github.com/openshift/origin/pkg/serviceaccounts/util"
+)
+
+const (
+	NewServiceAccountTokenRecommendedName = "new-token"
+
+	newServiceAccountTokenShort = `Generate a new token for a service account.`
+
+	newServiceAccountTokenLong = `
+Generate a new token for a service account.
+
+Service account API tokens are used by service accounts to authenticate to the API.
+This command will generate a new token, which could be used to compartmentalize service
+account actions by executing them with distinct tokens, to rotate out pre-existing token
+on the service account, or for use by an external client.
+`
+
+	newServiceAccountTokenUsage = `%s SA-NAME`
+
+	newServiceAccountTokenExamples = `  # Generate a new token and for service account 'custom-sa'
+  $ %[1]s 'custom-sa'
+`
+)
+
+type NewServiceAccountTokenOptions struct {
+	SAName        string
+	SAClient      unversioned.ServiceAccountsInterface
+	SecretsClient unversioned.SecretsInterface
+
+	Timeout time.Duration
+
+	Out io.Writer
+	Err io.Writer
+}
+
+func NewCommandNewServiceAccountToken(name, fullname string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
```

Comment:
Since the name is unpredictable, take an optional label to apply to the resources.  You should be able to use new-app or process as an example.  Otherwise, the user won't be able to know which secret is the one he made.